Add Dojo 2 as a Preset Template fixes #447

nicknisi opened the pull request:
<!--
Please make sure you are familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- Is it a Bug fix, feature, docs update, ... -->
**What kind of change does this PR introduce?**

This PR adds a new template for Dojo 2.

<!-- You can also link to an open issue here -->
**What is the current behavior?**
This is a new template, requested and described in #447 

<!-- if this is a feature change -->
**What is the new behavior?**

Dojo 2 template

<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [ ] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->

<!-- Thank you for contributing! -->
- add dojo2 template
- add temporary project as template to load (github)
- add type definition creation to styles transpiler
- add dojo 2 to home page
- add default loading of dojo template src/index.html and src/main.css

**Note:** This has a backend change in order to set the correct template value coming from the API. To work around this in development, I added the following line to packages/app/src/app/store/providers/Api.js:

```diff
diff --git i/packages/app/src/app/store/providers/Api.js w/packages/app/src/app/store/providers/Api.js
index 35248ccb..db410a04 100644
--- i/packages/app/src/app/store/providers/Api.js
+++ w/packages/app/src/app/store/providers/Api.js
@@ -78,6 +78,7 @@ function handleResponse(response, { shouldCamelize = true } = {}) {
     camelizedData.data.npmDependencies = response.result.data.npm_dependencies;
   }
 
+  camelizedData.data.template = '@dojo/cli-create-app';
   return camelizedData.data ? camelizedData.data : camelizedData;
 }
```

CompuIves commented on packages/app/src/app/components/CodeEditor/Monaco/index.js:
```diff
@@ -199,6 +202,18 @@ class MonacoEditor extends React.Component<Props, State> implements Editor {
   }
 
+  setupTranspilationListener() {
+    listen(({ type, code, path }) => {
```
This is reaaaally nice! I love that we now have the power of templates to define type definitions.

CompuIves commented on packages/app/src/sandbox/eval/transpilers/style/index.js:
```diff
@@ -24,12 +29,14 @@ class StyleTranspiler extends Transpiler {
         let result = insertCss(id, css);
         result += `\nmodule.exports=${JSON.stringify(exportTokens)};`;
 
+        dispatch({ type: 'add-extra-lib', path, code: classesToDefinition(exportTokens) });
```
I LOVE THIS EVEN MORE!

nicknisi commented:
@CompuIves Wow, thanks for the feedback! I went ahead and pushed up a commit with the API override for now so you should be able to test on staging.

I would be happy to document my learning and some of the difficulties! I'll start writing down my thoughts after work tonight.

CompuIves commented:
I made some changes in `master`, I'm sorry for that! I changed how Icons are resolved to utilize better code splitting. Could you rebase your branch on master? I'm planning to merge this in next Thursday.

nicknisi commented:
@CompuIves I rebased off the latest `master` and removed `DojoIcon` from the dojo template. That seemed to be all I needed to do to get it to work again! Let me know if there's any other way I can help!

Also, I am still working on a write-up about adding a new template. I should have it by this weekend at the latest.

dylans commented:
@CompuIves any update on when this might land? We're very close to releasing Dojo 2 and would like to mention this in the blog post announcing the Dojo 2 release. Thanks!

CompuIves commented:
@dylans I will have more time for CodeSandbox from tomorrow. So I'm either planning on releasing it today or tomorrow, sorry for the delay!

CompuIves commented:
Whoop! It's in now!
